File: ios/Packages/GemstonePrimitives/Sources/Formatters/FormattedNumber.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Formatters
import Foundation
import struct Gemstone.GemFormattedNumber
import enum Gemstone.GemNumberDisplay
import enum Gemstone.GemNumberNotation
import enum Gemstone.GemNumberRounding
import enum Gemstone.GemNumberUnit
import enum Gemstone.GemPrecision

public extension GemFormattedNumber {
    func text(locale: Locale = .current) -> String {
        switch notation {
        case .parenthesised: "(\(body(locale: locale)))"
        case .plain, .signed: body(locale: locale)
        }
    }
}

// MARK: - Private

private extension GemFormattedNumber {
    func body(locale: Locale) -> String {
        switch display {
        case let .number(precision):
            appendingSymbol(numberText(precision: precision, locale: locale))
        case .abbreviated:
            abbreviatedText(locale: locale)
        case let .belowThreshold(threshold, places):
            appendingSymbol("\(signText)<\(thresholdText(threshold, places: places, locale: locale))")
        }
    }

    var currencyCode: String? {
        switch unit {
        case let .currency(code): code
        case .percent, .symbol, .plain, .multiplier: nil
        }
    }

    var symbol: String? {
        switch unit {
        case let .symbol(symbol): symbol
        case .currency, .percent, .plain, .multiplier: nil
        }
    }

    var isPercent: Bool {
        switch unit {
        case .percent: true
        case .currency, .symbol, .plain, .multiplier: false
        }
    }

    var signText: String {
        guard showsSign else { return "" }
        return value < 0 ? "-" : "+"
    }

    var showsSign: Bool {
        switch notation {
        case .signed: true
        case .plain, .parenthesised: false
        }
    }

    var roundingRule: FloatingPointRoundingRule {
        switch rounding {
        case .toNearest: .toNearestOrEven
        case .towardZero: .towardZero
        }
    }

    var numberSign: NumberFormatStyleConfiguration.SignDisplayStrategy {
        showsSign ? .always(includingZero: true) : .automatic
    }

    var currencySign: CurrencyFormatStyleConfiguration.SignDisplayStrategy {
        showsSign ? .always(showZero: true) : .automatic
    }

    func numberText(precision: GemPrecision, locale: Locale) -> String {
        if isPercent {
            return value.formatted(
                .percent.locale(locale)
                    .precision(precision.formatStyle)
                    .sign(strategy: showsSign ? .always(includingZero: true) : .never)
                    .scale(1)
                    .rounded(rule: roundingRule),
            )
        }
        guard let currencyCode else {
            return value.formatted(.number.locale(locale).precision(precision.formatStyle).sign(strategy: numberSign).rounded(rule: roundingRule))
        }
        return value.formatted(.currency(code: currencyCode).locale(locale).precision(precision.formatStyle).sign(strategy: currencySign).rounded(rule: roundingRule))
    }

    func abbreviatedText(locale: Locale) -> String {
        let formatter = AbbreviatedFormatter(locale: locale)
        if let currencyCode {
            return formatter.string(from: value, currency: currencyCode, rule: roundingRule) ?? numberText(precision: .fraction(min: 2, max: 2), locale: locale)
        }
        return appendingSymbol(formatter.string(from: value, rule: roundingRule) ?? numberText(precision: .fraction(min: 2, max: 2), locale: locale))
    }

    func thresholdText(_ threshold: Double, places: UInt32, locale: Locale) -> String {
        guard let currencyCode else {
            return threshold.formatted(.number.locale(locale).precision(.fractionLength(Int(places))))
        }
        return threshold.formatted(.currency(code: currencyCode).locale(locale).precision(.fractionLength(Int(places))))
    }

    func appendingSymbol(_ text: String) -> String {
        if case .multiplier = unit {
            return "\(text)x"
        }
        guard let symbol else { return text }
        return "\(text) \(symbol)"
    }
}

File: android/gemcore/src/main/kotlin/com/gemwallet/android/model/FormattedNumber.kt
package com.gemwallet.android.model

import android.icu.text.CompactDecimalFormat
import uniffi.gemstone.GemFormattedNumber
import uniffi.gemstone.GemNumberDisplay
import uniffi.gemstone.GemNumberNotation
import uniffi.gemstone.GemNumberRounding
import uniffi.gemstone.GemNumberUnit
import uniffi.gemstone.GemPrecision
import uniffi.gemstone.GemTransactionRowValue
import java.math.BigDecimal
import java.math.RoundingMode
import java.text.DecimalFormat
import java.text.NumberFormat
import java.util.Locale

fun GemFormattedNumber.text(locale: Locale = Locale.getDefault()): String = when (notation) {
    GemNumberNotation.PARENTHESISED -> "(${body(locale)})"
    GemNumberNotation.PLAIN, GemNumberNotation.SIGNED -> body(locale)
}

private val GemFormattedNumber.showsSign: Boolean
    get() = notation == GemNumberNotation.SIGNED

private val GemFormattedNumber.numberRounding: RoundingMode
    get() = when (rounding) {
        GemNumberRounding.TO_NEAREST -> RoundingMode.HALF_EVEN
        GemNumberRounding.TOWARD_ZERO -> RoundingMode.DOWN
    }

private fun GemFormattedNumber.body(locale: Locale): String = when (val display = display) {
    is GemNumberDisplay.Number -> when (unit) {
        is GemNumberUnit.Percent -> percentText(BigDecimal.valueOf(value), display.precision, showsSign, numberRounding, locale)
        else -> appendSymbol(numberText(BigDecimal.valueOf(value), display.precision, locale))
    }

    is GemNumberDisplay.Abbreviated -> appendSymbol(abbreviatedText(BigDecimal.valueOf(value), locale))

    is GemNumberDisplay.BelowThreshold -> appendSymbol(
        "$signText<${numberText(BigDecimal.valueOf(display.threshold), GemPrecision.Fraction(display.places, display.places), locale, withSign = false)}",
    )
}

private val GemFormattedNumber.signText: String
    get() = when {
        !showsSign -> ""
        value < 0 -> "-"
        else -> "+"
    }

private val GemFormattedNumber.currencyCode: String?
    get() = (unit as? GemNumberUnit.Currency)?.code

private val GemFormattedNumber.symbol: String?
    get() = (unit as? GemNumberUnit.Symbol)?.symbol

private fun percentText(value: BigDecimal, precision: GemPrecision, showsSign: Boolean, rounding: RoundingMode, locale: Locale): String {
    val formatter = (NumberFormat.getPercentInstance(locale) as DecimalFormat).apply {
        when (precision) {
            is GemPrecision.Fraction -> {
                minimumFractionDigits = precision.min.toInt()
                maximumFractionDigits = precision.max.toInt()
            }

            is GemPrecision.Significant -> {
                minimumFractionDigits = 0
                maximumFractionDigits = Int.MAX_VALUE
            }
        }
        roundingMode = rounding
        val minus = decimalFormatSymbols.minusSign.toString()
        if (showsSign) {
            positivePrefix = "+" + positivePrefix
        } else {
            negativePrefix = negativePrefix.replace(minus, "")
        }
    }
    val amount = value.movePointLeft(2)
    return formatter.format(
        when (precision) {
            is GemPrecision.Fraction -> amount
            is GemPrecision.Significant -> amount.rounded(precision, rounding)
        },
    )
}

private fun GemFormattedNumber.appendSymbol(text: String): String = when (val unit = unit) {
    is GemNumberUnit.Symbol -> "$text ${unit.symbol}"
    GemNumberUnit.Multiplier -> "${text}x"
    is GemNumberUnit.Currency, GemNumberUnit.Percent, GemNumberUnit.Plain -> text
}

private fun GemFormattedNumber.numberText(value: BigDecimal, precision: GemPrecision, locale: Locale, withSign: Boolean = showsSign): String {
    val rounding = numberRounding
    val formatter = (numberFormat(locale) as DecimalFormat).apply {
        roundingMode = rounding
        if (withSign) {
            positivePrefix = "+" + positivePrefix
        }
    }
    return formatter.format(value, precision)
}

private fun GemFormattedNumber.abbreviatedText(value: BigDecimal, locale: Locale): String {
    val formatter = CompactDecimalFormat.getInstance(locale, CompactDecimalFormat.CompactStyle.SHORT).apply {
        setSignificantDigitsUsed(false)
        minimumFractionDigits = 0
        maximumFractionDigits = 2
        roundingMode = when (rounding) {
            GemNumberRounding.TO_NEAREST -> android.icu.math.BigDecimal.ROUND_HALF_EVEN
            GemNumberRounding.TOWARD_ZERO -> android.icu.math.BigDecimal.ROUND_DOWN
        }
        currencyCode?.let { currency = android.icu.util.Currency.getInstance(it) }
    }
    return when {
        showsSign -> "$signText${formatter.format(value.abs())}"
        else -> formatter.format(value)
    }
}

private fun GemFormattedNumber.numberFormat(locale: Locale): NumberFormat = currencyCode?.let { code ->
    NumberFormat.getCurrencyInstance(locale).apply { currency = java.util.Currency.getInstance(code) }
} ?: NumberFormat.getInstance(locale)

fun GemTransactionRowValue.text(): String? = when (this) {
    GemTransactionRowValue.None -> null
    is GemTransactionRowValue.AssetSymbol -> asset.symbol
    is GemTransactionRowValue.Number -> number.text()
}
